### PR TITLE
Obsolete SuppressIldasmAttribute and remove ildasm.exe support for it

### LINKED ISSUE #13341: Consider ignoring SuppressIldasmAttribute in ILDASM
The `SuppressIldasmAttribute` instructs the ildasm tool to prevent working on assemblies or modules that have this attribute [over here][1].

I'd like to ask re-considering if this is necessary still and if the check can be removed, and documenting the attribute as obsolete.

Some motivation:

The attribute is, as many know, trivial to bypass. Practically every other disassembly tool ignores this attribute. As for why "Well why not keep using that tool" is because ildasm offers some benefits over other tools, such as quickly dumping the IL-text to disk and re-assembling.

The attribute conveys as sense of protection that is really not existent. If the desired outcome is to protect intellectual property, then they should be steered toward 3rd party solutions like a commercial obfuscator.

[1]: https://github.com/dotnet/coreclr/blob/5f0c2f35ec943fcc7a360f059af1fc5b2b9ce1f5/src/ildasm/dasm.cpp#L881-L894

### LINKED ISSUE #40622: Obsolete SuppressIldasmAttribute
__Edit by @GrabYourPitchforks__: This issue tracks only obsoleting the attribute, not any other ildasm.exe work.

```cs
namespace System.Runtime.CompilerServices
{
    [Obsolete("SuppressIldasmAttribute has no effect in .NET 6.0+ applications.")] // NEW attribute
    [AttributeUsage(AttributeTargets.Assembly | AttributeTargets.Module)]
    public sealed class SuppressIldasmAttribute : Attribute
    {
        public SuppressIldasmAttribute() { }
    }
}
```

(What text do we even put there? "This attribute never made any sense to begin with and has no valid use case."? I've based the text proposed above on the `DisablePrivateReflectionAttribute` obsoletion text already in 6.0.)

-----

It has no valid use case, the security against reverse engineering it provides is nonexistent and there are tons of tools that ignore it.

My proposal is to deprecate it per #33360 and modify ILDasm to stop honoring that attribute.

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Resolves https://github.com/dotnet/runtime/issues/13341.
Resolves https://github.com/dotnet/runtime/issues/40622.

The obsoletion hasn't been officially approved yet, but based on email conversations we expect approval is imminent (maybe with some wording changes). I'm opening this PR optimistically to get a head start on CI runs and make sure there's no functional regression.

Temporarily marked as __no merge__ until API signoff is finalized.

@BruceForstall Do you anticipate any problem with me deleting the existing resource string and reassigning the resource ids to be gapless? I didn't see any functional regressions but wanted to pick your brain on this. If this is risky, I can restore the original resource ids and leave the gap.

### CODE REVIEW

**Comment by BruceForstall** [comment]
cc @jkotas in case you have opinions here

**Comment by GrabYourPitchforks** [comment]
There was some email discussion about what hardcoded message we should put in the attribute. IMO we should keep the hardcoded message short and sweet, and if we want to give further information on "here's why you can't rely on it for IP protection," that belongs in the aka.ms link that accompanies these attributes.

File: docs/project/list-of-obsoletions.md [hunk 1]
```diff
@@ -26,4 +26,5 @@ Currently the identifiers `SYSLIB0001` through `SYSLIB0999` are carved out for o
 |  __`SYSLIB0011`__ | `BinaryFormatter` serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for recommended alternatives. |
 |  __`SYSLIB0012`__ | Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location instead. |
 |  __`SYSLIB0013`__ | Uri.EscapeUriString can corrupt the Uri string in some cases. Consider using Uri.EscapeDataString for query string components instead. |
-|  __`SYSLIB0015`__ | DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications. |
\ No newline at end of file
+|  __`SYSLIB0015`__ | DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications. |
+|  __`SYSLIB0016`__ | SuppressIldasmAttribute has no effect in .NET 6.0+. |
```
Comment (stephentoub) [inline docs/project/list-of-obsoletions.md:30]: While you're here, should we change the message for SYSLIB0015 to be consistent, e.g. drop "applications"?

**Comment by briansull** [comment]
I see some ilasm testing failures:

```
  Starting:    Regressions.coreclr.XUnitWrapper
    Regressions/coreclr/GitHub_49982/test49982/test49982.sh [FAIL]
      usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
             cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
      rm: /private/tmp/helix/working/AF640960/w/AF7D096D/e/Regressions/coreclr/GitHub_49982/test49982/composite-r2r.dll.rsp: No such file or directory
      Unhandled exception. System.IO.FileLoadException: Could not load file or assembly '/private/tmp/helix/working/AF640960/w/AF7D096D/e/Regressions/coreclr/GitHub_49982/test49982/test49982.asm.dll'. The located assembly's manifest definition does not match the assembly reference. (0x80131040 (FUSION_E_REF_DEF_MISMATCH))
```

**Review by BruceForstall — APPROVED** [review]
ildasm changes LGTM

> @BruceForstall Do you anticipate any problem with me deleting the existing resource string and reassigning the resource ids to be gapless? I didn't see any functional regressions but wanted to pick your brain on this. If this is risky, I can restore the original resource ids and leave the gap.

I can't think of any reason why we would need to worry about this (but I don't have a *lot* of experience in this code)

**Comment by GrabYourPitchforks** [comment]
The failure reported by Brian (full logs [here](https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-50951-merge-2cb24ae9e9f743a495/PayloadGroup0/console.feb5f1e4.log?sv=2019-07-07&se=2021-04-28T21%3A52%3A52Z&sr=c&sp=rl&sig=PjDNBpLwdsS8TkQEOzTcZmPb62Trup78wbI9wNfPyLE%3D)) seems related to https://github.com/dotnet/runtime/pull/50364. That's a newly-introduced crossgen regression test. But I don't know why this PR would've introduced a regression in that code path. @trylek, thoughts?

I'm rerunning the test just in case it's flaky.

**Comment by trylek** [comment]
Hmm, I must admit this is the first time I'm seeing an ilasm test failure. This was certainly not the purpose of the new test, I suppose I should just mark it as ilasm-incompatible for now until the problem is understood; can someone please give me some pointers on how to run ilasm testing locally so that I can investigate what's wrong with the test? I guess that most likely the problem is that this is a "forced Crossgen2" test and if the test does something like ildasm followed by ilasm, applying that to a R2R executable certainly won't produce an identical result.

**Comment by BruceForstall** [comment]
To run the round-trip testing, set `RunningIlasmRoundTrip=1` before executing the test wrapper script. Note there are other environment variables set, e.g., `COMPlus_TieredCompilation=0`

**Comment by trylek** [comment]
Thanks Bruce for clarifying! It can easily be the case that it's just about wrong ordering of the steps in the <code>CLRTest.*.targets</code> scripts under <code>src\tests\Common</code> - if we make sure to run the ildasm - ilasm roundtrip before executing Crossgen2 (or Crossgen1, for that matter), I bet it would work just fine.

File: docs/project/list-of-obsoletions.md [hunk 2]
```diff
@@ -26,4 +26,5 @@ Currently the identifiers `SYSLIB0001` through `SYSLIB0999` are carved out for o
 |  __`SYSLIB0011`__ | `BinaryFormatter` serialization is obsolete and should not be used. See https://aka.ms/binaryformatter for recommended alternatives. |
 |  __`SYSLIB0012`__ | Assembly.CodeBase and Assembly.EscapedCodeBase are only included for .NET Framework compatibility. Use Assembly.Location instead. |
 |  __`SYSLIB0013`__ | Uri.EscapeUriString can corrupt the Uri string in some cases. Consider using Uri.EscapeDataString for query string components instead. |
-|  __`SYSLIB0015`__ | DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications. |
\ No newline at end of file
+|  __`SYSLIB0015`__ | DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications. |
```
Comment (jeffhandley) [inline docs/project/list-of-obsoletions.md:29]:
```suggestion
|  __`SYSLIB0015`__ | DisablePrivateReflectionAttribute has no effect in .NET 6.0+. |
```

File: src/libraries/Common/src/System/Obsoletions.cs [hunk 3]
```diff
@@ -51,5 +51,8 @@ internal static class Obsoletions
 
         internal const string DisablePrivateReflectionAttributeMessage = "DisablePrivateReflectionAttribute has no effect in .NET 6.0+ applications.";
```
Comment (jeffhandley) [inline src/libraries/Common/src/System/Obsoletions.cs:52]:
```suggestion
        internal const string DisablePrivateReflectionAttributeMessage = "DisablePrivateReflectionAttribute has no effect in .NET 6.0+.";
```

**Comment by jeffhandley** [comment]
SYSLIB0016 has been claimed by another obsoletion. I'm going to update this to be SYSLIB0024.

**Review by jeffhandley — APPROVED** [review]

**Comment by GrabYourPitchforks** [comment]
I'll resolve the conflicts and get this in shortly.

**Comment by GrabYourPitchforks** [comment]
Changes since Jeff's latest iteration:
- Bumped diagnostic code to __SYSLIB0025__.
- Fixed a stray whitespace issue in the .builds file.

Otherwise, no functional changes from previous iteration. Since PR is already signed off I'll merge once CI goes green.

**Comment by GrabYourPitchforks** [comment]
Test failure is known issue https://github.com/dotnet/runtime/issues/48851.

**Review by jeffhandley — APPROVED** [review]

**Comment by jeffhandley** [comment]
Breaking change documentation will be handled by dotnet/docs#23856. /cc @gewarren

**Pull request merged**